File: starters/py/src/Game.py
'''
Created on 28 juil. 2013

@author: Mike
'''
from math import ceil, sqrt
from sys import stdout, stderr

class Game(object):
    '''
    Class handling communication to the DTStrike server ans providing the basics methods to play.
    '''

    MY_ID = 1

    def __init__(self, data):
        """ Parse options and send go when ready to rock and roll ! """
        stdout.write("go\n")
        stdout.flush()
    
    def start_turn(self, data):
        self.planets = []
        self.fleets = []
        self.parse_game_state(data)
    
    def num_ships(self, playerID):
        '''
        Returns the number of ships that the playerID has, either located
        on planets or in flight.
        '''
        num_ships = 0
        for planet in self.planets:
            if planet.owner == playerID:
                num_ships += planet.num_ships            
        for fleet in self.fleets:
            if fleet.owner == playerID:
                num_ships += fleet.num_ships()
        return num_ships
    
    def is_alive(self, playerID):
        '''
        A player is alive if he owns at least one military planet or one fleet.
        '''
        for planet in self.planets:
            if(isinstance(planet, MilitaryPlanet) & planet.owner == playerID):
                return True
        for fleet in self.fleets:
            if fleet.owner == playerID:
                return True
        return False

    def my_planets(self):
        '''
        Return a list of all the planets owned by the current player. By
        convention, the current player is always player number 1.
        '''
        return [p for p in self.planets if p.owner == self.MY_ID]
    
    def my_economic_planets(self):
        '''
        Return a list of all the economic planets owned by the current player. By
        convention, the current player is always player number 1.
        '''        
        return [p for p in self.planets if p.owner == self.MY_ID & isinstance(p, EconomicPlanet)]


    def neutral_economic_planets(self):
        '''
        Return a list of all neutral economic planets.
        '''
        return [p for p in self.planets if p.owner == 0 & isinstance(p, EconomicPlanet)]

    def enemy_economic_planets(self):
        '''
        Return a list of all the economic planets owned by rival players. This excludes
        economic planets owned by the current player, as well as neutral economic planets.
        '''
        return [p for p in self.planets if p.owner > self.MY_ID & isinstance(p, EconomicPlanet)]
        
    def my_military_fleets(self):
        '''
        Return a list of all the military fleets owned by the current player. By
        convention, the current player is always player number 1.
        '''
        return [f for f in self.fleets if f.owner == 1 & f.military]

    def my_military_planets(self):
        '''
        Return a list of all the military fleets owned by the current player. By
        convention, the current player is always player number 1.
        '''
        return [p for p in self.planets if p.owner == 1 & isinstance(p, MilitaryPlanet)]
           
    def military_planets(self):
        '''
        Return a list of all the military planets.
        '''
        return [p for p in self.planets if isinstance(p, MilitaryPlanet)]

    def neutral_military_planets(self):
        '''
        Return a list of all neutral military planets.
        '''
        return [p for p in self.planets if p.owner == 0 & isinstance(p, MilitaryPlanet)]
    
    def enemy_military_planets(self):
        '''
        Return a list of all the military planets owned by rival players. This excludes
        military planets owned by the current player, as well as neutral military planets.
        '''        
        return [p for p in self.planets if p.owner > self.MY_ID & isinstance(p, MilitaryPlanet)]

    def neutral_planets(self):
        '''
        Return a list of all neutral planets.
        '''
        return [p for p in self.planets if p.owner == 0]

    def enemy_planets(self):
        '''
        Return a list of all the planets owned by rival players. This excludes
        planets owned by the current player, as well as neutral planets.
        '''
        return [p for p in self.planets if p.owner > self.MY_ID]
    
    def not_my_planets(self):
        '''
        Return a list of all the planets that are not owned by the current
        player. This includes all enemy planets and neutral planets.
        '''
        return [p for p in self.planets if p.owner != self.MY_ID]

    def not_my_fleets(self):
        '''
        Return a list of all the fleets not owned by the current player.
        '''
        return [f for f in self.fleets if f.owner != self.MY_ID]

    def my_fleets(self):
        '''
        Return a list of all the fleets owned by the current player.
        '''
        return [f for f in self.fleets if f.owner == self.MY_ID]
    
    def enemy_fleets(self):
        '''
        Return a list of all the fleets owned by enemy players.
        '''
        return [f for f in self.fleets if f.owner > self.MY_ID]

    def distance(self, source_planet_id, destination_planet_id):
        source = self.planets[source_planet_id]
        destination = self.planets[destination_planet_id]
        dx = source.X() - destination.X()
        dy = source.Y() - destination.Y()
        return ceil(sqrt(dx * dx + dy * dy))
    
    def issue_order(self, source_planet_id, destination_planet_id, num_ships):
        '''
        Sends an order to the game engine. An order is composed of a source
        planet number, a destination planet number, and a number of ships. A
        few things to keep in mind:
        * you can issue many orders per turn if you like.
        * the planets are numbered starting at zero, not one.
        * you must own the source planet. If you break this rule, the game
        engine kicks your bot out of the game instantly.
        * you can't move more ships than are currently on the source planet.
        * the ships will take a few turns to reach their destination. Travel
        is not instant. See the distance() function for more info.
        '''
        stdout.write("%d %d %d\n" % (source_planet_id, destination_planet_id, num_ships))
        stdout.flush()
        
    def finish_turn(self):
        '''
        Sends the game engine a message to let it know that we're done sending
         orders. This means the end of our turn.
        '''
        stdout.write("go\n")
        stdout.flush()
        
    def end_game(self, data):
        ### do nothing
        stderr.write("bye")
    
    def parse_game_state(self, data):
        for line in data:
            line = line.split("#")[0] # remove comments?!
            tokens = line.split(" ")
            if (len(tokens) > 1): #remove empty strings
                if (tokens[0] == "M"):
                    if (len(tokens) != 5):
                        return 1
                    p = MilitaryPlanet(len(self.planets), # ID of this planet
                       int(tokens[3]), # Owner
                       int(tokens[4]), # Num ships
                       float(tokens[1]), # X
                       float(tokens[2])) # Y
                    self.planets.append(p)
                elif tokens[0] == "E":
                    if len(tokens) != 6:
                        return 1
                    p = EconomicPlanet(len(self.planets), # ID of this planet
                       int(tokens[3]), # Owner
                       int(tokens[4]), # Num ships
                       float(tokens[1]), # X
                       float(tokens[2]), # Y
                       int(tokens[5])) # Income
                    self.planets.append(p)
                elif tokens[0] == "F" or tokens[0] == "R":
                    if len(tokens) != 7:
                        return 1
                    f = Fleet(int(tokens[1]), # Owner
                      int(tokens[2]), # Num ships
                      int(tokens[3]), # Source
                      int(tokens[4]), # Destination
                      int(tokens[5]), # Total trip length
                      int(tokens[6]), # Turns remaining
                      tokens[0] == "F") # Military fleet
                    self.fleets.append(f)


class Fleet(object):
    '''
    classdocs
    '''
    def __init__(self, owner, power, sourceDept = -1, destDept = -1, tripLength = -1, turnsRemaining = -1, military = True):
        '''
        Constructor
        '''
        self.owner = owner
        self.num_ships = power
        self.sourcePlanet = sourceDept
        self.destinationPlanet = destDept
        self.totalTripLength = tripLength
        self.turnsRemaining = turnsRemaining
        self.military = military
        
class Planet(object):
    '''
    A Planet, mother class.
    '''
    def __init__(self, id, owner, num_ships, x, y):
        '''
        Constructor
        '''
        self.id = id
        self.owner = owner
        self.num_ships = num_ships
        self.x = x
        self.y = y
        
        
class EconomicPlanet(Planet):
    '''
    Eco planet with income
    '''
    def __init__(self, id, owner, num_ships, x, y, income):
        '''
        Constructor
        '''
        Planet.__init__(self, id, owner, num_ships, x, y)
        self.income = income        
        
class MilitaryPlanet(Planet):
    '''
    Military Planet, which can handle ships.
    '''
    def __init__(self, id, owner, num_ships, x, y):
        '''
        Constructor
        '''
        Planet.__init__(self, id, owner, num_ships, x, y)

File: starters/c#/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Globalization;
using System.Text;
using System.Collections;
using System.IO;

namespace DTStrike.MyBot
{
    public class Game
    {
        public List<Planet> planets;
	    public List<Fleet> fleets;
        public const int MY_ID = 1;

	    public Game(List<String> data) {
		    planets = new List<Planet>();
		    fleets = new List<Fleet>();
		    if (parse(data) > 0) {
                System.Console.Error.WriteLine("erreur parse...");
		    }
	    }

	    /**
	     * A player is alive if he owns at least one military planet or one fleet.
	     * 
	     * @param id
	     *            player id to check
	     * @return true if the player is alive
	     */
	    public bool isAlive(int id) {
		    foreach (Planet p in planets) {
			    if (p is MilitaryPlanet && p.owner == id)
				    return true;
		    }
		    foreach (Fleet f in fleets) {
			    if (f.owner == id && f.militaryFleet) {
				    return true;
			    }
		    }
		    return false;
	    }

	// Returns the number of planets. Planets are numbered starting with 0.
	public int numPlanets() {
		return planets.Count();
	}

	// Returns the planet with the given planet_id. There are NumPlanets()
	// planets. They are numbered starting at 0.
	public Planet getPlanet(int planetID) {
		return planets.Find(x=>x.id == planetID);
	}

	// Returns the number of fleets.
	public int numFleets() {
		return fleets.Count();
	}

	// Returns the fleet with the given fleet_id. Fleets are numbered starting
	// with 0. There are NumFleets() fleets. fleet_id's are not consistent from
	// one turn to the next.
	public Fleet getFleet(int fleetID) {
		return fleets.Find(x=>x.owner == fleetID);
	}

	// Returns a list of all the planets.
	public List<Planet> getPlanets() {
		return planets;
	}

	// Return a list of all the planets owned by the current player. By
	// convention, the current player is always player number 1.
	public List<Planet> getMyPlanets() {
		List<Planet> r = new List<Planet>();
		foreach (Planet p in planets) {
			if (p.owner == MY_ID) {
				r.Add(p);
			}
		}
		return r;
	}

	// Return a list of all the planets owned by the current player. By
	// convention, the current player is always player number 1.
	public List<Planet> getMyMilitaryPlanets() {
		List<Planet> r = new List<Planet>();
		foreach (Planet p in planets) {
			if (p.owner == MY_ID && p is MilitaryPlanet) {
				r.Add(p);
			}
		}
		return r;
	}

	// Return a list of all neutral planets.
	public List<Planet> getNeutralPlanets() {
		List<Planet> r = new List<Planet>();
		foreach (Planet p in planets) {
			if (p.owner == 0) {
				r.Add(p);
			}
		}
		return r;
	}

	// Return a list of all the planets owned by rival players. This excludes
	// planets owned by the current player, as well as neutral planets.
	public List<Planet> getEnemyPlanets() {
		List<Planet> r = new List<Planet>();
		foreach (Planet p in planets) {
			if (p.owner != 0 && p.owner != MY_ID) {
				r.Add(p);
			}
		}
		return r;
	}

	// Return a list of all the planets that are not owned by the current
	// player. This includes all enemy planets and neutral planets.
	public List<Planet> getNotMyPlanets() {
		List<Planet> r = new List<Planet>();
		foreach (Planet p in planets) {
			if (p.owner != MY_ID) {
				r.Add(p);
			}
		}
		return r;
	}

	// Return a list of all the fleets.
	public List<Fleet> getFleets() {
		List<Fleet> r = new List<Fleet>();
		foreach (Fleet f in fleets) {
			r.Add(f);
		}
		return r;
	}

	// Return a list of all the fleets owned by the current player.
	public List<Fleet> getMyFleets() {
		List<Fleet> r = new List<Fleet>();
		foreach (Fleet f in fleets) {
			if (f.owner == MY_ID) {
				r.Add(f);
			}
		}
		return r;
	}

    // Return a list of all the fleets owned by the current player.
    public List<Fleet> getMyMilitaryFleets()
    {
        List<Fleet> r = new List<Fleet>();
        foreach (Fleet f in fleets)
        {
            if (f.owner == MY_ID && f.militaryFleet)
            {
                r.Add(f);
            }
        }
        return r;
    }

	// Return a list of all the fleets owned by enemy players.
	public List<Fleet> getEnemyFleets() {
		List<Fleet> r = new List<Fleet>();
		foreach (Fleet f in fleets) {
			if (f.owner != MY_ID) {
				r.Add(f);
			}
		}
		return r;
	}

	// Returns the distance between two planets, rounded up to the next highest
	// integer. This is the number of discrete time steps it takes to get
	// between the two planets.
	public int distance(int sourcePlanet, int destinationPlanet) {
		Planet source = planets.Find(x=>x.id == sourcePlanet);
		Planet destination = planets.Find(x=>x.id == destinationPlanet);
		double dx = source.x - destination.x;
		double dy = source.y - destination.y;
		return (int) Math.Ceiling(Math.Sqrt(dx * dx + dy * dy));
	}

	// Sends an order to the game engine. An order is composed of a source
	// planet number, a destination planet number, and a number of ships. A
	// few things to keep in mind:
	// * you can issue many orders per turn if you like.
	// * the planets are numbered starting at zero, not one.
	// * you must own the source planet. If you break this rule, the game
	// engine kicks your bot out of the game instantly.
	// * you can't move more ships than are currently on the source planet.
	// * the ships will take a few turns to reach their destination. Travel
	// is not instant. See the Distance() function for more info.
	public void issueOrder(int sourcePlanet, int destinationPlanet, int numShips) {
		System.Console.Out.Write("" + sourcePlanet + " " + destinationPlanet + " "
				+ numShips + "\n");
        System.Console.Out.Flush();
	}

	// Sends an order to the game engine. An order is composed of a source
	// planet number, a destination planet number, and a number of ships. A
	// few things to keep in mind:
	// * you can issue many orders per turn if you like.
	// * the planets are numbered starting at zero, not one.
	// * you must own the source planet. If you break this rule, the game
	// engine kicks your bot out of the game instantly.
	// * you can't move more ships than are currently on the source planet.
	// * the ships will take a few turns to reach their destination. Travel
	// is not instant. See the Distance() function for more info.
	public void issueOrder(Planet source, Planet dest, int numShips) {
		System.Console.Out.Write("" + source.id + " " + dest.id + " " + numShips + "\n");
        System.Console.Out.Flush();
	}

	// Sends the game engine a message to let it know that we're done sending
	// orders. This signifies the end of our turn.
	public void finishTurn() {
        System.Console.Out.Write("go\n");
        System.Console.Out.Flush();
	}

	// Returns the number of ships that the current player has, either located
	// on planets or in flight.
	public int numShips(int playerID) {
		int numShips = 0;
		foreach (Planet p in planets) {
			if (p.owner == playerID) {
				numShips += p.numShips;
			}
		}
		foreach (Fleet f in fleets) {
			if (f.owner == playerID) {
				numShips += f.numShips;
			}
		}
		return numShips;
	}

	private int parse(List<String> data) {
		planets = new List<Planet>();
        fleets = new List<Fleet>();

        CultureInfo en = CultureInfo.GetCultureInfo("en-US"); // Doubles are written with a . separator

		foreach (String str in data) {
			if (str.Trim().Length != 0) {
				String[] line = str.Split(' ');
				if (line.Length != 0) {
					if (line[0].Equals("M")) {
						if (line.Length != 5) {
							System.Console.Error.WriteLine("error line 0: " + str);
							return 1;
						}
						double x = Double.Parse(line[1], en);
                        double y = Double.Parse(line[2], en);
                        int owner = int.Parse(line[3]);
                        int numShips = int.Parse(line[4]);
                        
						MilitaryPlanet p = new MilitaryPlanet(planets.Count(),
								owner, numShips, x, y);
						planets.Add(p);
					} else if (line[0].Equals("E")) {
						if (line.Length != 6) {
							System.Console.Error.WriteLine("error line 1: " + str);
							return 1;
						}
                        double x = Double.Parse(line[1], en);
                        double y = Double.Parse(line[2], en);
                        int owner = int.Parse(line[3]);
                        int numShips = int.Parse(line[4]);
						int economicValue = int.Parse(line[5]);

						EconomicPlanet p = new EconomicPlanet(planets.Count(),
								owner, numShips, economicValue, x, y);
						planets.Add(p);
					} else if (line[0].Equals("F") || line[0].Equals("R")) {
						if (line.Length != 7) {
							System.Console.Error.WriteLine("error line 2: " + str);
							return 1;
						}
                        int owner = int.Parse(line[1]);
                        int numShips = int.Parse(line[2]);
                        int sourceDept = int.Parse(line[3]);
                        int destDept = int.Parse(line[4]);
                        int tripLength = int.Parse(line[5]);
                        int turnsRemaining = int.Parse(line[6]);
						Fleet fleet = new Fleet(owner, numShips, sourceDept,
                                destDept, tripLength, turnsRemaining, line[0].Equals("F"));
						this.fleets.Add(fleet);
					} else if (line[0].Equals("turn")) {
						// New turn
					}
				}
			}
		}
		return 0;
	}
    }
}
